custom_components/versatile_thermostat/underlyings.py

```diff
@@ -247,9 +245,14 @@ def is_device_active(self):
             not self.is_inversed and real_state
         )
 
+    async def _keep_alive_callback(self):
+        """Keep alive: Turn on if already turned on, turn off if already turned off."""
+        await (self.turn_on() if self.is_device_active else self.turn_off())
+
     # @overrides this breaks some unit tests TypeError: object MagicMock can't be used in 'await' expression
     async def turn_off(self):
         """Turn heater toggleable device off."""
+        self._keep_alive.cancel()  # Cancel early to avoid a turn_on/turn_off race condition
```

Comment:
This `cancel()` call here and similarly in the `turn_on()` method is something that I had added to my original keep-alive PR (#345), and you pushed back saying that it was _“Not necessary because `set_async_action` already cancels”_ ([link](https://github.com/jmcollin78/versatile_thermostat/pull/345#discussion_r1459450453)). That‘s true, but I had pointed out ([link](https://github.com/jmcollin78/versatile_thermostat/pull/345#discussion_r1460058632)) that there is an async call (`await` statement) between this `cancel()` call and the `set_async_action()` call, which causes an async context switch during which other queued async tasks get a chance to run. An imminent keep-alive callback could be among those other queued tasks, and theoretically a `turn_on` callback could be called in the middle of a `turn_off` method. In the original PR, I had ultimately reasoned that even if this happened, the correct service call should immediately follow, so I treated this early `cancel()` call as an optimisation that was not sufficiently justified and thus I agreed to delete it.

However, I cannot “prove” that preventing a `turn_on` callback from being called in the middle of the `turn_off` method is ”just an optimisation” rather than a ”correctness requirement,” and therefore I prefer to make this early `cancel()` call to simply avoid this situation in the first place — especially in the context of suspicions (#379). If you can prove that the early `cancel()` call is not required, then we can delete it again. 🧐😀

Having said that, the user (#379) claimed that their issue was (consistently?) reproducible and my reasoning above should only apply to ”rare corner cases.” Still, I’d rather fend off against rare corner cases as well.